### ISSUE TITLE
LEMUR training/indexing mismatch with batch-scope centering

### ISSUE THREAD

**Issue by morgan-coded** [neuml/txtai#1190]
I found an interaction between #1168's batch-centering behavior and the LEMUR trainer that I didn't account for when these changes landed. Batch-composition dependence itself was intentional and documented in #1168. The narrower issue is that LEMUR training and indexing apply different centering transforms under the same configuration.

The trainer encodes every text with `batch=1` (`pipeline/train/lemur.py` — the loop at `documents = [pooling.encode([text], batch=1, ...)`), so under scope `batch` the centering mean degenerates to each document's own tokens. Indexing encodes `encodebatch` (default 32) documents per model batch and centers across all of them. The same document can therefore receive different centered token embeddings depending on the path and the other documents sharing its encoding batch.

Reproduced at `3ea19746` with `neuml/colbert-bert-tiny`, `center: true`, 64 synthetic documents (script below):

- Token vectors, trainer-vs-index: maximum absolute coordinate difference 0.305; mean per-document maximum absolute coordinate difference 0.164. The `scope: document` control removes the discrepancy under the same padding path (≤ 7e-7), isolating batch-scope centering as the cause in this reproduction. Deterministic across reruns.
- Exact MaxSim over the two encodings: the ordered top-5 list differed for 8/8 synthetic queries. This does not imply relevance quality declined.
- End-to-end with an `epochs=0` artifact trained through `LemurTrainer`: LEMUR feature-vector maximum absolute coordinate difference 0.036; the ordered top-5 list differed for 8/8 queries; the `scope: document` LEMUR control is ≤ 2e-7.

This demonstrates path inconsistency; it does not establish which resulting ranking has better retrieval quality.

Is batch-scoped centering intended to be supported for LEMUR artifacts? If so, what invariant should training, initial indexing, and later updates share? If not, should LEMUR require a batch-invariant scope such as `document`? I held off on proposing a patch because that contract determines which path should change.

<details><summary>Reproduction script</summary>

```python
# Centering-parity reproduction - txtai @ 3ea197468, neuml/colbert-bert-tiny, CPU
import random
import tempfile

import numpy as np
import torch

from txtai.models import Models, PoolingFactory
from txtai.pipeline import LemurTrainer

MODEL = "neuml/colbert-bert-tiny"
BATCH = 32  # txtai encodebatch default
random.seed(42)
VOCAB = (
    "retrieval index vector token model corpus query document ranking search "
    "neural embedding scoring latency memory training feature centering batch "
    "apple river mountain quantum ledger orbit fabric signal harbor lattice"
).split()
DOCS = [" ".join(random.choices(VOCAB, k=random.randint(4, 60))) for _ in range(64)]
QUERIES = [" ".join(random.choices(VOCAB, k=random.randint(3, 8))) for _ in range(8)]


def pool(center, lemur=None):
    # mirrors LemurTrainer's pooling construction
    return PoolingFactory.create(
        {"method": None, "path": MODEL, "device": Models.deviceid(False), "tokenizer": None,
         "maxlength": None, "modelargs": {"center": center, "muvera": None, "lemur": lemur}}
    )


def asnp(x):
    return x.detach().cpu().numpy() if isinstance(x, torch.Tensor) else np.asarray(x)


def trainer_style(p, texts, category="data"):  # batch=1, as pipeline/train/lemur.py
    return [asnp(p.encode([t], batch=1, category=category)[0]) for t in texts]


def index_style(p, texts, category="data"):  # encodebatch default
    return [asnp(v) for v in p.encode(texts, batch=BATCH, category=category)]


def trim(a, b):  # batch encoding pads token rows; real tokens are the leading rows
    return b[: a.shape[0]] if b.ndim == a.ndim and b.shape[0] > a.shape[0] else b


def report(name, A, B):
    diffs = [np.abs(a - trim(a, b)).max() for a, b in zip(A, B)]
    print(f"[{name}] max={max(diffs):.6g} mean_per_doc_max={np.mean(diffs):.6g}")


def maxsim_top5(q, docs):
    return list(np.argsort([float((q @ d.T).max(axis=1).sum()) for d in docs])[::-1][:5])


def fixed_top5(q, docs):
    return list(np.argsort([float(np.dot(np.ravel(q), np.ravel(d))) for d in docs])[::-1][:5])


pb = pool(center=True)
print("resolved center:", pb.center)  # {'scope': 'batch'}
A, B = trainer_style(pb, DOCS), index_style(pb, DOCS)
report("batch-scope trainer-vs-index", A, B)
Q = trainer_style(pb, QUERIES, category="query")
flips = sum(maxsim_top5(q, A) != maxsim_top5(q, [trim(a, b) for a, b in zip(A, B)]) for q in Q)
print(f"top-5 changed for {flips}/{len(Q)} queries")

pd = pool(center={"scope": "document"})
report("document-scope control", trainer_style(pd, DOCS), index_style(pd, DOCS))

out = tempfile.mkdtemp(prefix="lemur-parity-")
LemurTrainer()(MODEL, DOCS, out, gpu=False, vectors={"center": True}, epochs=0,
               trainsubsetsize=64, learnsubsetsize=2000, olssamplesize=512)
pl = pool(center=True, lemur={"path": out})
LA, LB = trainer_style(pl, DOCS), index_style(pl, DOCS)
report("LEMUR fixed vectors batch-scope", LA, LB)
QL = trainer_style(pl, QUERIES, category="query")
flips = sum(fixed_top5(q, LA) != fixed_top5(q, LB) for q in QL)
print(f"LEMUR top-5 changed for {flips}/{len(QL)} queries")

pld = pool(center={"scope": "document"}, lemur={"path": out})
report("LEMUR document-scope control", trainer_style(pld, DOCS), index_style(pld, DOCS))
```

</details>

**Comment by davidmezzetti** [comment]
In general the advantage of `batch` mode is that you don't have to do any preprocessing like `collection`. With document while always deterministic, the results weren't as good.

As long as a dataset is processed in a deterministic order, wouldn't `batch` mode always give the same results? If we shuffle sort or otherwise change the order, it will change how centering works. 

Is this just a caveat of `batch` mode rather than it being "broken" though?

EDIT/IDEA: Are you suggesting we run collection mode centering with LEMUR training and save those weights with the model? Then use that centering as the default if a LEMUR model is available (otherwise it falls back to batch or selected method)?

**Comment by morgan-coded** [comment]
Right — with a fixed order and pipeline, batch mode is fully deterministic: the repro's rerun was byte-identical, and length-sorted batching brought even a shuffled corpus back to float noise (7e-8). Composition sensitivity on its own is the documented #1168 caveat, agreed.

The part I'd call more than a caveat is train-vs-index specifically: the trainer encodes every text at `batch=1`, so under `batch` scope the artifact is fitted on per-document-centered vectors, while indexing encodes through 32-document batches — the same document gets two geometries under one config. That's what the 0.036 fixed-vector divergence and the 8/8 top-5 changes measure, with order held fixed. A side effect is that batch mode's measured quality may even be a touch understated — the artifacts in any comparison were fitted on geometry the index path doesn't produce.

Your EDIT is exactly what I'd do: compute the collection mean during LEMUR training, save it with the artifact, and auto-select `collection` centering with that mean whenever a LEMUR model is present, falling back to `batch`/selected otherwise. Training and indexing then center identically by construction, batch keeps its no-preprocessing advantage everywhere else, and half the plumbing exists — `collection` already takes a provided mean. Happy to build it that way if you want.

**Comment by davidmezzetti** [comment]
Since we're doing a pass over the data when training LEMUR it makes total sense to utilize that to also factor in mean centering. If that isn't available the current `batch` scope still make the most sense as the default to me.
